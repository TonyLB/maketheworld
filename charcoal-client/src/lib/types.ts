export type NarrowOmit<T, K extends keyof any> = T extends any ? Omit<T, K> : never

export const assertTypeguard = <T extends any, G extends T>(value: T, typeguard: (value) => value is G): G | undefined => {
    if (value && typeguard(value)) {
        return value
    }
    return undefined
}
