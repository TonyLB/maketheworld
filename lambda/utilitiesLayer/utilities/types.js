export const splitType = (value) => {
    if (value) {
        const sections = value.split('#')
        if (sections.length) {
            return [sections[0], sections.slice(1).join('#')]
        }
    }
    return ['', '']
}
