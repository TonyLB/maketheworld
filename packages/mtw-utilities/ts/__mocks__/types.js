export const splitType = (value) => {
    if (value) {
        const sections = value.split('#')
        if (sections.length) {
            return [sections[0], sections.slice(1).join('#')]
        }
    }
    return ['', '']
}

export const AssetKey = (assetId) => (`ASSET#${assetId}`)
export const CharacterKey = (characterId) => (`CHARACTER#${characterId}`)
export const RoomKey = (roomId) => (`ROOM#${roomId}`)
