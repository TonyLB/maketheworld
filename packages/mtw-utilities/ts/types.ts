export const splitType = (value: string) => {
    if (value) {
        const sections = value.split('#')
        if (sections.length) {
            return [sections[0], sections.slice(1).join('#')]
        }
    }
    return ['', '']
}

export const AssetKey = (assetId: string) => (`ASSET#${assetId}`)
export const CharacterKey = (characterId: string) => (`CHARACTER#${characterId}`)
export const RoomKey = (roomId: string) => (`ROOM#${roomId}`)
