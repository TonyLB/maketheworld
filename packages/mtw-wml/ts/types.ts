export type ArrayContents<T extends Array<any>> = T extends Array<infer G> ? G : never
