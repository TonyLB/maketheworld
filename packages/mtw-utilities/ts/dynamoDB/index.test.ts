
var mockDBClientSend = jest.fn()
var mockDBClient = jest.fn(() => ({
    send: mockDBClientSend
}))
var mockGetItemCommand = jest.fn()
var mockUpdateItemCommand = jest.fn()
var mockPutItemCommand = jest.fn()
var mockDeleteItemCommand = jest.fn()
var mockQueryCommand = jest.fn()
var mockTransactWriteItemsCommand = jest.fn()

jest.mock('@aws-sdk/client-dynamodb', () => ({
    DynamoDBClient: mockDBClient,
    GetItemCommand: mockGetItemCommand,
    UpdateItemCommand: mockUpdateItemCommand,
    PutItemCommand: mockPutItemCommand,
    DeleteItemCommand: mockDeleteItemCommand,
    QueryCommand: mockQueryCommand,
    BatchWriteItemCommand: jest.fn(),
    BatchGetItemCommand: jest.fn(),
    TransactWriteItemsCommand: mockTransactWriteItemsCommand
}))

import { marshall } from '@aws-sdk/util-dynamodb'
import { abstractOptimisticUpdate, addPerAsset } from '.'
import { splitType } from '../types'

describe('optimisticUpdate', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
    })

    it('should do nothing on no change of state', async () => {
        mockDBClientSend.mockResolvedValue({ Item: marshall({ test: 'Testing' }) })
        const returnValue = await abstractOptimisticUpdate('ephemeraTest')({
            key: {
                EphemeraId: 'TEST',
                DataCategory: 'Meta::Test'
            },
            updateKeys: ['test'],
            updateReducer: () => {}
        })
        expect(returnValue).toEqual({ test: 'Testing' })
        expect(mockUpdateItemCommand).not.toHaveBeenCalled()
    })

    it('should update and remove fields on change', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({ Item: marshall({ testOne: 'Testing', testTwo: 'Also Testing', testFour: 'Unchanged' }) })
            .mockResolvedValueOnce({ Attributes: marshall({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', testOne: 'Different Test', testThree: 'New test', testFour: 'Unchanged' })})
        const returnValue = await abstractOptimisticUpdate('ephemeraTest')({
            key: {
                EphemeraId: 'TEST',
                DataCategory: 'Meta::Test'
            },
            updateKeys: ['testOne', 'testTwo', 'testThree', 'testFour'],
            updateReducer: (draft) => {
                draft.testOne = 'Different Test',
                draft.testTwo = undefined
                draft.testThree = 'New test'
            }
        })
        expect(returnValue).toEqual({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', testOne: 'Different Test', testThree: 'New test', testFour: 'Unchanged' })
        expect(mockUpdateItemCommand).toHaveBeenCalledTimes(1)
        expect(mockUpdateItemCommand.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should update when no item fetched', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({})
            .mockResolvedValueOnce({ Attributes: marshall({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', testOne: 'Test', testTwo: 'Another test' })})
        const returnValue = await abstractOptimisticUpdate('ephemeraTest')({
            key: {
                EphemeraId: 'TEST',
                DataCategory: 'Meta::Test'
            },
            updateKeys: ['testOne', 'testTwo'],
            updateReducer: (draft) => {
                draft.testOne = 'Test'
                draft.testTwo = 'Another test'
            }
        })
        expect(returnValue).toEqual({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', testOne: 'Test', testTwo: 'Another test' })
        expect(mockUpdateItemCommand).toHaveBeenCalledTimes(1)
        expect(mockUpdateItemCommand.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should update when field not defined', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({ Item: marshall({ EphemeraId: 'TEST', DataCategory: 'Meta::Test' })})
            .mockResolvedValueOnce({ Attributes: marshall({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', optionalField: 'Test' })})
        const returnValue = await abstractOptimisticUpdate('ephemeraTest')({
            key: {
                EphemeraId: 'TEST',
                DataCategory: 'Meta::Test'
            },
            updateKeys: ['optionalField'],
            updateReducer: (draft) => {
                draft.optionalField = 'Test'
            }
        })
        expect(returnValue).toEqual({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', optionalField: 'Test' })
        expect(mockUpdateItemCommand).toHaveBeenCalledTimes(1)
        expect(mockUpdateItemCommand.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should properly map ExpressionAttributeNames', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({ Item: marshall({ Name: 'Testing', Test: 'Also Testing', testFour: 'Unchanged' }) })
            .mockResolvedValueOnce({ Attributes: marshall({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', Name: 'Different Test', Zone: 'New test', testFour: 'Unchanged' })})
        const returnValue = await abstractOptimisticUpdate('ephemeraTest')({
            key: {
                EphemeraId: 'TEST',
                DataCategory: 'Meta::Test'
            },
            ExpressionAttributeNames: {
                '#name': 'Name',
                '#zone': 'Zone',
                '#test': 'Test'
            },
            updateKeys: ['#name', '#test', '#zone', 'testFour'],
            updateReducer: (draft) => {
                draft.Name = 'Different Test',
                draft.Test = undefined
                draft.Zone = 'New test'
            }
        })
        expect(returnValue).toEqual({ EphemeraId: 'TEST', DataCategory: 'Meta::Test', Name: 'Different Test', Zone: 'New test', testFour: 'Unchanged' })
        expect(mockUpdateItemCommand).toHaveBeenCalledTimes(1)
        expect(mockUpdateItemCommand.mock.calls[0][0]).toMatchSnapshot()
    })

})

type TestAddPerAssetArgs = {
    EphemeraId: string;
    DataCategory: string;
    value: string;
}

describe('addPerAsset', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
    })

    it('should create a new Meta record', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({})

        await addPerAsset({
            updateKeys: ['cached'],
            reduceMetaData: ({ item }: { item: TestAddPerAssetArgs }) => (draft) => {
                const key = splitType(item.EphemeraId)[1]
                if (!draft.cached) {
                    draft.cached = []
                }
                if (!(key in draft.cached)) {
                    draft.cached.push(key)
                }
            }
        })({
            EphemeraId: 'FEATURE#TEST',
            DataCategory: 'ASSET#TEST',
            value: 'Test'
        })
        expect(mockTransactWriteItemsCommand).toHaveBeenCalledTimes(1)
        expect(mockTransactWriteItemsCommand.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should update an existing Meta record', async () => {
        mockDBClientSend
            .mockResolvedValueOnce({ Item: marshall({ cached: ['OLD'] }) })
        await addPerAsset({
            updateKeys: ['cached'],
            reduceMetaData: ({ item }: { item: TestAddPerAssetArgs }) => (draft) => {
                const key = splitType(item.EphemeraId)[1]
                if (!draft.cached) {
                    draft.cached = []
                }
                if (!(key in draft.cached)) {
                    draft.cached.push(key)
                }
            }
        })({
            EphemeraId: 'FEATURE#TEST',
            DataCategory: 'ASSET#TEST',
            value: 'Test'
        })
        expect(mockTransactWriteItemsCommand).toHaveBeenCalledTimes(1)
        expect(mockTransactWriteItemsCommand.mock.calls[0][0]).toMatchSnapshot()
    })

})