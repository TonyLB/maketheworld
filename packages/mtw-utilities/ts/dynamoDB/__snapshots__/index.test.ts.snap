// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPerAsset should create a new Meta record 1`] = `
Object {
  "TransactItems": Array [
    Object {
      "Put": Object {
        "Item": Object {
          "DataCategory": Object {
            "S": "ASSET#TEST",
          },
          "EphemeraId": Object {
            "S": "FEATURE#TEST",
          },
          "value": Object {
            "S": "Test",
          },
        },
        "TableName": "undefined_ephemera",
      },
    },
    Object {
      "Update": Object {
        "ConditionExpression": "attribute_not_exists(DataCategory)",
        "ExpressionAttributeValues": Object {
          ":New0": Object {
            "L": Array [
              Object {
                "S": "TEST",
              },
            ],
          },
        },
        "Key": Object {
          "DataCategory": Object {
            "S": "Meta::Feature",
          },
          "EphemeraId": Object {
            "S": "FEATURE#TEST",
          },
        },
        "TableName": "undefined_ephemera",
        "UpdateExpression": "SET cached = :New0",
      },
    },
  ],
}
`;

exports[`addPerAsset should update an existing Meta record 1`] = `
Object {
  "TransactItems": Array [
    Object {
      "Put": Object {
        "Item": Object {
          "DataCategory": Object {
            "S": "ASSET#TEST",
          },
          "EphemeraId": Object {
            "S": "FEATURE#TEST",
          },
          "value": Object {
            "S": "Test",
          },
        },
        "TableName": "undefined_ephemera",
      },
    },
    Object {
      "Update": Object {
        "ConditionExpression": "cached = :Old0",
        "ExpressionAttributeValues": Object {
          ":New0": Object {
            "L": Array [
              Object {
                "S": "OLD",
              },
              Object {
                "S": "TEST",
              },
            ],
          },
          ":Old0": Object {
            "L": Array [
              Object {
                "S": "OLD",
              },
            ],
          },
        },
        "Key": Object {
          "DataCategory": Object {
            "S": "Meta::Feature",
          },
          "EphemeraId": Object {
            "S": "FEATURE#TEST",
          },
        },
        "TableName": "undefined_ephemera",
        "UpdateExpression": "SET cached = :New0",
      },
    },
  ],
}
`;

exports[`optimisticUpdate should properly map ExpressionAttributeNames 1`] = `
Object {
  "ConditionExpression": "#name = :Old0 AND #test = :Old1 AND attribute_not_exists(#zone)",
  "ExpressionAttributeNames": Object {
    "#name": "Name",
    "#test": "Test",
    "#zone": "Zone",
  },
  "ExpressionAttributeValues": Object {
    ":New0": Object {
      "S": "Different Test",
    },
    ":New2": Object {
      "S": "New test",
    },
    ":Old0": Object {
      "S": "Testing",
    },
    ":Old1": Object {
      "S": "Also Testing",
    },
  },
  "Key": Object {
    "DataCategory": Object {
      "S": "Meta::Test",
    },
    "EphemeraId": Object {
      "S": "TEST",
    },
  },
  "ReturnValues": undefined,
  "TableName": "ephemeraTest",
  "UpdateExpression": "SET #name = :New0, #zone = :New2 REMOVE #test",
}
`;

exports[`optimisticUpdate should update and remove fields on change 1`] = `
Object {
  "ConditionExpression": "testOne = :Old0 AND testTwo = :Old1 AND attribute_not_exists(testThree)",
  "ExpressionAttributeValues": Object {
    ":New0": Object {
      "S": "Different Test",
    },
    ":New2": Object {
      "S": "New test",
    },
    ":Old0": Object {
      "S": "Testing",
    },
    ":Old1": Object {
      "S": "Also Testing",
    },
  },
  "Key": Object {
    "DataCategory": Object {
      "S": "Meta::Test",
    },
    "EphemeraId": Object {
      "S": "TEST",
    },
  },
  "ReturnValues": undefined,
  "TableName": "ephemeraTest",
  "UpdateExpression": "SET testOne = :New0, testThree = :New2 REMOVE testTwo",
}
`;

exports[`optimisticUpdate should update when field not defined 1`] = `
Object {
  "ConditionExpression": "attribute_not_exists(optionalField)",
  "ExpressionAttributeValues": Object {
    ":New0": Object {
      "S": "Test",
    },
  },
  "Key": Object {
    "DataCategory": Object {
      "S": "Meta::Test",
    },
    "EphemeraId": Object {
      "S": "TEST",
    },
  },
  "ReturnValues": undefined,
  "TableName": "ephemeraTest",
  "UpdateExpression": "SET optionalField = :New0",
}
`;

exports[`optimisticUpdate should update when no item fetched 1`] = `
Object {
  "ConditionExpression": "attribute_not_exists(DataCategory)",
  "ExpressionAttributeValues": Object {
    ":New0": Object {
      "S": "Test",
    },
    ":New1": Object {
      "S": "Another test",
    },
  },
  "Key": Object {
    "DataCategory": Object {
      "S": "Meta::Test",
    },
    "EphemeraId": Object {
      "S": "TEST",
    },
  },
  "ReturnValues": undefined,
  "TableName": "ephemeraTest",
  "UpdateExpression": "SET testOne = :New0, testTwo = :New1",
}
`;
